### PR TITLE
haskell: pkgsStatic fixes

### PR DESCRIPTION
**Pull request by rnhmjoj**
###### Motivation for this change

Fixing common build failures in pkgsStatic.haskellPackages

See the commit messages for details.

###### Things done

- built static GHCs 9.0.2 and 9.2.1
- built `pkgsStatic.haskellPackages.{QuickCheck,pretty-simple}` (previously failing)
- built and tested `pkgsStatic.haskellPackages.pandoc`

### CODE REVIEW

File: pkgs/development/haskell-modules/generic-builder.nix
Comment (sternenseemann) [inline pkgs/development/haskell-modules/generic-builder.nix:59]: I think a better should would be to add an `hasHaddock` or similar passthru arg to `ghc` and use this here.

**Comment by sternenseemann** [comment]
* I'd drop d8dcb588aed526a632c3060c1b091908c7cf31df, since it changes the meaning of `haskell.packages.ghc*` / `haskellPackages` between `pkgs` and `pkgsStatic` which I find problematic. Especially since the compiler build itself is not affected, this doesn't seem necessary?
* Can you add a package to the `staticHaskellPackages` jobset in `pkgs/top-level/release-haskell.nix` which doesn't build without `-fexternal-dynamic-refs`? This would help keeping tabs on it.

**Comment by rnhmjoj** [comment]
> Especially since the compiler build itself is not affected, this doesn't seem necessary?

I see your point, though I'm not sure how useful the default compiler is, given most packages fails to build. Ideally, we should fix the linking errors, but avoiding gmp is the only way I've found so far.

> Can you add a package to the staticHaskellPackages jobset in pkgs/top-level/release-haskell.nix which doesn't build without -fexternal-dynamic-refs? This would help keeping tabs on it.

Done.

File: pkgs/development/compilers/ghc/9.2.1.nix
Comment (cdepillabout) [inline pkgs/development/compilers/ghc/9.2.1.nix:111]: Would you be able to add a comment here why `-fexternal-dynamic-refs` is needed for `enableRelocatedStaticLibs`?

From the GHC manual: https://downloads.haskell.org/ghc/latest/docs/html/users_guide/phases.html#ghc-flag--fexternal-dynamic-refs

> When generating code, assume that entities imported from a different module might be dynamically linked. This flag is enabled automatically by [-dynamic](https://downloads.haskell.org/ghc/latest/docs/html/users_guide/phases.html#ghc-flag--dynamic).

It is not clear to me why this would be necessary if we are statically linking things.  (Although I don't have a good understanding of static linking, so maybe it should be obvious.)

Comment (rnhmjoj) [reply pkgs/development/compilers/ghc/9.2.1.nix:111]: I have to admit I'm baffled by this too and was kind of expecting the haskell maintainers would know better.
Basically, I'm implementing the solution proposed in this blog post: https://www.tweag.io/blog/2020-09-30-bazel-static-haskell/.

Comment (cdepillabout) [reply pkgs/development/compilers/ghc/9.2.1.nix:111]: Thanks for the link to the blog post.  It says:

> When evaluating template Haskell splices, GHC will execute compiled expressions in its built-in bytecode interpreter and this code has to be compatible with the RTS of GHC itself. In short, a GHC configured with a dynamic RTS will not be able to load static Haskell libraries to evaluate template Haskell splices.

It then goes on to say:

> We need to build GHC with a static RTS and to make sure that Haskell code is compiled as position independent code so that it can be loaded into a running GHC for template Haskell splices.

Both of these statements make sense of course, but they don't seem to explain why `-fexternal-dynamic-refs` is necessary.

File: pkgs/development/compilers/ghc/9.0.2.nix
Comment (cdepillabout) [inline pkgs/development/compilers/ghc/9.0.2.nix:110]: Sorry, GitHub isn't letting me create an actual suggestion here, but how about writing this as:

```suggestion
  '' + 
  # These options are recommended by
  # https://www.tweag.io/blog/2020-09-30-bazel-static-haskell/
  # We're not sure why -fexternal-dynamic-refs is necessary.
  lib.optionalString enableRelocatedStaticLibs ''
    GhcLibHcOpts += -fPIC -fexternal-dynamic-refs
    GhcRtsHcOpts += -fPIC -fexternal-dynamic-refs
```

This same change should be made for `9.2.1.nix` as well.

**Review by cdepillabout — APPROVED** [review]
Aside from the above comment suggestions, this LGTM!

I tried `nix-build -A pkgsStatic.pretty-simple` and it appears to build and run correctly!

File: pkgs/development/compilers/ghc/9.0.2.nix
Comment (rnhmjoj) [reply pkgs/development/compilers/ghc/9.0.2.nix:110]: Thanks, I added something like this.

**Comment by cdepillabout** [comment]
@sternenseemann Does this look alright to merge to you?

File: pkgs/development/compilers/ghc/9.2.1.nix
Comment (sternenseemann) [reply pkgs/development/compilers/ghc/9.2.1.nix:115]: The GHC man page reads:

> Generate code for linking against dynamic libraries

I assume this means that the RTS will have some kind of dynamic library loading facilities even of static (Haskell?) archives, so the byte code interpreter / whatever used by TH will actually work. `-fPIC` also serves this purpose.

**Comment by rnhmjoj** [comment]
> but I would be fine with changing the meaning of top level haskellPackages as a compromise,

Looks great to me! Thanks for the suggestion.

File: pkgs/top-level/all-packages.nix
Comment (sternenseemann) [inline pkgs/top-level/all-packages.nix:12452]: The old `targetPackages.haskellPackages.ghc or …` needs to be preserved, since it's not the same as `haskell.compiler.ghc902` (e.g. it has a `withPackages` method).

**Review by sternenseemann — CHANGES_REQUESTED** [review]

File: pkgs/top-level/all-packages.nix
Comment (rnhmjoj) [reply pkgs/top-level/all-packages.nix:12452]: I messed up by copying one too many line from the other change. Thanks.

File: pkgs/development/haskell-modules/generic-builder.nix
Comment (cdepillabout) [reply pkgs/development/haskell-modules/generic-builder.nix:58]: I think this suggestion makes sense.  But on the other hand, as an end user, if I pass `doHaddock = true`, I'd expect either I get haddocks when the build finishes, or the build fails.  It could be somewhat confusing to specify `doHaddock = true`, but have the build successfully finish without building haddocks.

Either way, I'd say this potential fix probably shouldn't hold up merging in this PR.

**Comment by cdepillabout** [comment]
> I would be fine with changing the meaning of top level haskellPackages as a compromise

I agree this was a good suggestion.  I think the current state of this PR makes a lot of sense, having `pkgsStatic.haskellPackages` explicitly pointing at `pkgsStatic.haskell.packages.native-bignum.ghc-902`.

> Have you tried actually building anything that uses FFI in TH? I guess that would need a dynamically linked library as well?

That's a good question.  Do you happen to know a package that uses FFI in TH off the top of your head?  I gave it some thought, but I couldn't come up with anything.  (I did try compiling `pkgsStatic.haskellPackages.zlib`, which does successfully compile, but I imagine it is unlikely to use TH.)

Although even if this doesn't work, I don't feel like it should hold up merging this PR, since this PR currently improves the status quo.

**Comment by rnhmjoj** [comment]
So, can we merge this?

**Comment by sternenseemann** [comment]
Probably, I'll try to have a final look soon. A small request: Could you use `haskellPackages.mkDerivation` over `haskell-generic-builder` in the commit message? I'd like to have correspondence to an attribute where possible. Rebasing on haskell-updates would be useful anyways, so CI wouldn't be broken by an unrelated problem.

**Comment by rnhmjoj** [comment]
Commit message updated and rebased.

**Review by sternenseemann — CHANGES_REQUESTED** [review]

File: pkgs/top-level/all-packages.nix
Comment (rnhmjoj) [reply pkgs/top-level/all-packages.nix:12497]: It seems so. Fixed, again.

File: pkgs/development/haskell-modules/generic-builder.nix
Comment (sternenseemann) [inline pkgs/development/haskell-modules/generic-builder.nix:58]: The remaining GHCs (8.8.4, 8.10.7, HEAD) need this annotation as well, e. g. `ghc*Binary`, but also `ghcjs`, see the current eval error as per ofborg.

**Review by sternenseemann — CHANGES_REQUESTED** [review]

File: pkgs/development/haskell-modules/generic-builder.nix
Comment (rnhmjoj) [reply pkgs/development/haskell-modules/generic-builder.nix:58]: The binary compilers don't have the `enableHaddockProgram` arguments. What should I set `hasHaddock` to?

Comment (rnhmjoj) [reply pkgs/development/haskell-modules/generic-builder.nix:58]: What about just doing `&& (ghc.hasHaddock or true)`?

Comment (sternenseemann) [reply pkgs/development/haskell-modules/generic-builder.nix:58]: > What about just doing `&& (ghc.hasHaddock or true)`?

Also works and probably a smart move since it means that we don't have to commit to having this attribute forever.

Comment (rnhmjoj) [reply pkgs/development/haskell-modules/generic-builder.nix]: Ok, did it this way. I also added `hasHaddock` to `ghcHEAD`, so the main versions are covered.

Comment (rnhmjoj) [reply pkgs/development/haskell-modules/generic-builder.nix]: The evaluation is fixed, now.

Comment (sternenseemann) [reply pkgs/development/haskell-modules/generic-builder.nix]: GHC 8.8.4 and 8.10.7 also have an `enableHaddockProgram` argument, don't they?

Comment (rnhmjoj) [reply pkgs/development/haskell-modules/generic-builder.nix]: Yes, but I though it would be enough to target the next major version 9.0 and head.
Do you want me to add those too?

Comment (sternenseemann) [reply pkgs/development/haskell-modules/generic-builder.nix]: Yeah, definitely, the package sets are still around and work for the most part and they should behave like the default ones as much as possible!

Comment (rnhmjoj) [reply pkgs/development/haskell-modules/generic-builder.nix]: Done.

**Comment by rnhmjoj** [comment]
@sternenseemann Thank you for reviewing!

**Pull request merged**
